Fixed some game-breaking Lingo decompilation bugs

juanmv94 opened the pull request:
I did some bug fixes for the Lingo decompilation including:

* Operator precedence issues. Example: "(var1 +var2) / var3" was decompiled as "var1 +var2 / var3"
* Sprite "within" operation was reversed. Example: "sprite 1 within 2" was decompiled as "sprite 2 within 1"
* Globals on handlers were ignored and named as "unknown0". Globals were explored on the handler instead, but this was breaking some do("string evaluated code") calls.
* Logic like:
```
if (cond1) then
    ...
    repeat while (cond2)
        ...
    end repeat
end if
```

was decompiled as:
```
repeat while (cond1)
    ...
    repeat while (cond2)
        ...
    end repeat
end repeat
```

mathiaskivi commented:
Tested out the changes in your repository and these actually fixed the issues I had with Alias Runner (https://github.com/ProjectorRays/ProjectorRays/issues/24). 👍❤️
Supersonic RC main menu still breaks though.

emansom commented:
@Quackster!

juanmv94 commented:
I've just added 2 more fixes:
* After getting globals in my first commit, I've noticed that some DXRs might have unused globalIDs = -1 (I don't know the reason, and can't replicate). I've filtered out these globalIDs.
* I've found some "_-- ERROR: Could not identify jmp_" lines on some Lingo scripts that didn't worked, so I had a look and found the issue coming from "exit repeat" logic inside an "if" block being detected as an "else". This made something like:
```
on exitFrame
  repeat with x = 1 to 5
    if random(x)>random(x+1) then
      put("ok")
      exit repeat
    end if
    put("nook")
  end repeat
  put("end")
end
```
to be decompiled as:
```
on exitFrame
  repeat with x = 1 to 5
    if (random(x) > random((x + 1))) then
      put "ok"
    else
      put "nook"
    end if
  end repeat
  put "end"
end
```
I've fixed this problem moving the "exit repeat" and "next repeat" detection logic before "else" detection logic that was giving false positives.

emansom commented on src/director/handler.cpp:
```diff
@@ -130,7 +130,8 @@ void Handler::readNames() {
 	}
 	for (auto nameID : globalNameIDs) {
-		globalNames.push_back(getName(nameID));
+		if (nameID>=0)	// Some global nameIDs = -1 might exist
```
Needs curly braces here.

djsrv replied:
The `>=` operator should have spaces around it.

djsrv commented on src/director/handler.cpp:
```diff
@@ -376,7 +373,7 @@ void Handler::tagLoops() {
 		uint32_t endIndex = bytecodePosMap[jmpPos];
 		auto &endRepeat = bytecodeArray[endIndex - 1];
-		if (endRepeat.opcode != kOpEndRepeat)
+		if (endRepeat.opcode != kOpEndRepeat || (endRepeat.pos - endRepeat.obj)>jmpifz.pos)
```
The `>` operator should have spaces around it.

djsrv reviewed (changes requested):
Hi, thank you for the fixes!

I try to stick to atomic commits (i.e. one fix per commit) so that they're easier to understand and revert if later needed. These commits have multiple changes each so they should be broken apart.

I also use the ScummVM code formatting conventions so that code can be shared with ScummVM's Director engine: https://wiki.scummvm.org/index.php/Code_Formatting_Conventions There are a couple places where these conventions aren't followed which I've commented on.

If you're comfortable with command-line git, I'd appreciate if you [rewrote the commits](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) to fix those. If not, I can rewrite them when merging.

juanmv94 committed:
Getting globals from handlers

juanmv94 committed:
Fix if->repeat blocks being decompiled as repeat->repeat

juanmv94 committed:
Fix within and intersects sprite operations being reversed

juanmv94 committed:
Fix exit repeat to be detected as an else statement when on repeat->if block

juanmv94 committed:
Add workarround to fix operator precedence problems

juanmv94 commented:
All done @djsrv

djsrv commented:
Thank you
